pkg/pex/discovery.go

```diff
@@ -68,6 +67,10 @@ func (disc *discover) startTracking(ctx context.Context, log log.Logger, ns stri
 				t.Reset(disc.advertise(ctx, log, ns))
 
 			case <-ctx.Done():
+				disc.mu.Lock()
```

Comment:
You are right, nice catch! And indeed, this was not a cleaner design, but a bug that I found while debugging. Because if the context was cancelled, the `advertise` was not cleaned-up and so, consequent calls to advertise didn't take effect.